### PR TITLE
Add Guarded, Otherwise, Guarded1, Otherwise1

### PR DESCRIPTION
**Pull request by blmage**
I'm open to suggestions about those ones! (especially regarding the names or the signatures)

### CODE REVIEW

File: src/Fcf.hs [hunk 1]
```diff
@@ -302,6 +302,18 @@ data Not :: Bool -> Exp Bool
 type instance Eval (Not 'True)  = 'False
 type instance Eval (Not 'False) = 'True
 
+type Otherwise = 'True
+
+data Guarded :: [(Bool, Exp a)] -> Exp a
+type instance Eval (Guarded ('(p, x) ': xs)) =
+    If p (Eval x) (Eval (Guarded xs))
```
Comment (Lysxia) [inline src/Fcf.hs]: Commuting `Eval` and `If` (so `Eval` is on the outside) would make this lazier. Currently, if `p` and `Eval x` get stuck, `Eval (Guarded xs)` will be evaluated.

**Comment by Lysxia** [comment]
Thanks, that looks useful! I'll leave this open to think about bikeshedding until next week.

File: src/Fcf.hs (hunk 1 above)
Comment (blmage) [reply src/Fcf.hs]: I had overlooked this, thanks!

**Comment by blmage** [comment]
Thanks for all the quick reviews!

I'll make sure to send more PRs as I stumble upon new use cases which seem suitable enough :)

**Comment by Lysxia** [comment]
I'm wondering whether this is too specific or complex. 

```haskell
Guarded
  [ ( b1, x1 )
  , ( b2, x2 )
  , ( Otherwise, x3 )
  ]
```

is equivalent to

```haskell
If b1 x1 $
If b2 x2 $
        x3
```

One motivating use case might be to pass `Guarded` to higher-order functions, but that seems oddly specific.

Do you have examples where `Guarded` or `Guarded1` is preferable?

**Comment by blmage** [comment]
Sorry for the delay! I've just pushed the requested changes, let me know whether everything seems fine now.

**Comment by Lysxia** [comment]
That's great, thanks!

**Comment by Lysxia** [comment]
I'll adapt your example to make a little test case!

**Pull request merged**